### ISSUE TITLE
Material with alpha scissor does not cast a shadow

### ISSUE THREAD

**Issue by BastiaanOlij** [godotengine/godot#19444]
**Godot version:**
Godot 3 master commit c4dfef58d9cca9d9112f5f8d8db9dcc1bd9dfaae

**OS/device including version:**
Windows 10, but don't think it matters

**Issue description:**
I have a material for my grass, just a texture on a plane, that has an alpha scissor applied so I only render the blades of grass.
Without this turned on, my plane casts a nice rectangular shadow.
With this turned on, no shadow.
I'm expecting a shadow in the shape of the grass particles.

I think it is because its treating the material as a transparent material.

**Steps to reproduce:**
Create a plane, add a material with a texture and turn alpha scissor on.
(obviously, create something the shadow is cast on and create a light so there is a shadow)

**Comment by Zylann** [comment]
Duplicate of https://github.com/godotengine/godot/issues/13350?

**Comment by BastiaanOlij** [comment]
@zireael07 yup, that's my understanding of how it should work, but it may not be detecting it properly and ending up getting rendered in the transparency pass

**Comment by BastiaanOlij** [comment]
@Zylann could be related but setting render_mode to depth_draw_alpha_prepass does not make a difference for me.

**Comment by BastiaanOlij** [comment]
Made some progress on this, @reduz could use your input.

The main issue seems to be around this code:
```
#if defined(ALPHA_SCISSOR_USED)
	if (alpha<alpha_scissor) {
		discard;
	}
#endif

#ifdef USE_OPAQUE_PREPASS

	if (alpha<0.99) {
		discard;
	}
#endif
```
( https://github.com/godotengine/godot/blob/master/drivers/gles3/shaders/scene.glsl#L1676 )

The alpha scissor nicely discards the fragment at the correct scissor value,  but for the shadow to be rendered during the opaque prepass a hardcoded value of 0.99 is used. 

In my example the grass' alpha is causing most of the fragments to be discarded.

As a workaround I'm applying the scissor directly in my fragment shader and avoiding the use of alpha:
```
	if (color.a < 0.3) {
		discard;
	}
```

I think the solution would be to do something like this:

```
#if defined(ALPHA_SCISSOR_USED)
	if (alpha<alpha_scissor) {
		discard;
	}
#else
#ifdef USE_OPAQUE_PREPASS

	if (alpha<0.99) {
		discard;
	}
#endif
#endif
```

Does the shader support nested ifdefs? or have an #elifdef ?

**Comment by Calinou** [comment]
> Does the shader support nested ifdefs? or have an #elifdef ?

Apparently, yes. I've seen something like this in `scene.glsl` for instance:

```glsl
#ifdef SHADOW_MODE_PCF_13
    ...
#elif defined(SHADOW_MODE_PCF_5)
    ...
#else
    ...
#endif
```

**Comment by BastiaanOlij** [comment]
I think this was fixed by ad8897984183d29b52128d8ad5db6ca7d2ff8278
